app/src/play/java/com/amaze/filemanager/utils/PackageInstallValidation.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2014-2022 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
+ *
+ * This file is part of Amaze File Manager.
+ *
+ * Amaze File Manager is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.amaze.filemanager.utils
+
+import android.content.pm.PackageInfo
+import android.content.pm.PackageManager
+import com.amaze.filemanager.application.AppConfig
+import com.amaze.filemanager.filesystem.files.FileUtils
+import java.io.File
+
+/**
+ * Additional checks for package validity for installation via Amaze.
+ *
+ */
+object PackageInstallValidation {
+
+    /**
+     * Perform validation by getting [PackageInfo] of file at specified path, then see if it's us.
+     * If yes, throw [PackageCannotBeInstalledException] and [FileUtils.installApk] should exit
+     * with Toast.
+     */
+    @JvmStatic
+    @Throws(PackageCannotBeInstalledException::class)
+    fun validatePackageInstallability(f: File) {
+        AppConfig.getInstance().run {
+            val packageInfo: PackageInfo? =
+                packageManager.getPackageArchiveInfo(
+                    f.absolutePath,
+                    PackageManager.GET_ACTIVITIES
+                )
+            if (packageInfo == null) {
+                throw PackageCannotBeInstalledException("Cannot get package info")
```

Comment:
both cases are handled in the same way, is this intentional? i.e, "Cannot get package info" & "Cannot update myself per Google Play policy" shows the Toast message that the app can't be installed due to Google's limitations. a workaround that comes to my mind is to throw `IllegalStateException` like you did before & handle it with a different Toast message.